lint/structLitKeyOrder_checker.go

```diff
@@ -0,0 +1,75 @@
+package lint
+
+import (
+	"go/ast"
+	"go/types"
+
+	"github.com/go-critic/go-critic/lint/internal/lintutil"
+	"github.com/go-toolsmith/astp"
+)
+
+func init() {
+	addChecker(&structLitKeyOrderChecker{}, attrVeryOpinionated, attrExperimental)
+}
+
+type structLitKeyOrderChecker struct {
+	checkerBase
+}
+
+func (c *structLitKeyOrderChecker) InitDocumentation(d *Documentation) {
+	d.Summary = "Detects struct literal keys order that does not match declaration order"
+	d.Before = `
+type foo struct{ x, y int }
+v := foo{y: y, x: x}`
+	d.After = `
+type foo struct{ x, y int }
+v := foo{x: x, y: y}`
+}
+
+func (c *structLitKeyOrderChecker) VisitExpr(expr ast.Expr) {
+	lit, ok := expr.(*ast.CompositeLit)
+	if !ok || len(lit.Elts) <= 1 {
+		// If there is 1 or less elements, consider the
+		// list as sorted. This also makes addressing
+		// the first element below safe.
+		return
+	}
+	if !astp.IsKeyValueExpr(lit.Elts[0]) {
+		// Skip literals without keyed field initializers.
+		// If at least 1 field is initialized using keyed
+		// syntax, all of them are guaranteed to be
+		// initialized by a KeyValueExpr.
+		return
+	}
+	typ, ok := c.ctx.typesInfo.TypeOf(lit).Underlying().(*types.Struct)
+	if !ok {
+		return
+	}
+
+	// Collect field name to declaration order mapping.
+	// There are other solutions that do not involve allocated
+	// map, but the current implementation is fast enough.
+	fieldsOrder := make(map[string]int)
+	for i := 0; i < typ.NumFields(); i++ {
+		fieldsOrder[typ.Field(i).Name()] = i
+	}
+
+	// Check until we find a field that is misplaced.
+	// Misplaced field is a field that comes after a field
+	// that had higher declaration order value.
+	maxOrder := -1
+	for _, elt := range lit.Elts {
+		kv := elt.(*ast.KeyValueExpr)
+		order := fieldsOrder[lintutil.AsIdent(kv.Key).Name]
+		if order > maxOrder {
```

Comment:
I think this `if` can be inverted easily 😉

Reply:
I don't get it, unfortunately.
Can you show me the code?
For me it seems like inverting here does not provide any improvement.

Reply:
Yeah, sorry for that. I mean fast exit:
```
if order <= maxOrder {
	c.warn(expr)
	return
}
maxOrder = order
```
but it's about style mostly 😉